dimod/include/dimod/adjarraybqm.h

```diff
@@ -89,6 +89,131 @@ class AdjArrayBQM {
+        size_type* counters_cumsum = (size_type*)utils::aligned_calloc(num_variables, sizeof(size_type));
+
+        // Backup copy of the diagonal of the dense matrix.
+        std::vector<B2> dense_diagonal(num_variables);
+
+        #pragma omp parallel
+        {
+            // Zero out the diagonal to avoid expensive checks inside innermost
+            // loop in the code for reading the matrix. The diagonal will be
+            // restored so a backup copy is saved.
+            #pragma omp for schedule(static)
+            for (size_type u = 0; u < num_variables; ++u) {
+                dense_diagonal[u] = dense[u * (num_variables + 1)];
+                dense[u * (num_variables + 1)] = 0;
+            }
+
+            // We process the matrix in two passes, in the first pass we take note of
+            // the number of total elements and elements in each row for proper memory
+            // allocation. In the second pass we fill up our desired bqm. We process
+            // the matrix in blocks of size BLOCK_SIZE*BLOCK_SIZE to take advantage of
+            // cache locality. Dynamic scheduling is used as we know some blocks may
+            // be more sparse than others and processing them may finish earlier.
+            #pragma omp for schedule(dynamic)
+            for (size_type u_st = 0; u_st < num_variables; u_st += BLOCK_SIZE) {
+                size_type u_end = std::min(u_st + BLOCK_SIZE, num_variables);
+                for (size_type v_st = 0; v_st < num_variables; v_st += BLOCK_SIZE) {
+                    size_type v_end = std::min(v_st + BLOCK_SIZE, num_variables);
+                    for (size_type u = u_st; u < u_end; ++u) {
+                        size_type counter_u = counters_cumsum[u];
+                        size_type counter_u_old = counter_u;
+                        for (size_type v = v_st; v < v_end; ++v) {
+                            bias_type qbias = dense[u * num_variables + v] + dense[v * num_variables + u];
+                            if (qbias != 0) {
+                                counter_u++;
+                            }
+                        }
+                        if (counter_u != counter_u_old) {
+                            counters_cumsum[u] = counter_u;
+                        }
+                    }
+                }
+            }
+        }
+
+        // Calculate the exclusive scan of the counters. Thus it will contain the starting
+        // indices in the bqm for inserting elements for each row of the dense matrix. This
+        // could be parallelized but has diminishing returns.
+        size_type sum_counters = 0;
+        for (size_type u = 0; u < num_variables; ++u) {
+            size_type prev_counter = counters_cumsum[u];
+            counters_cumsum[u] = sum_counters;
+            sum_counters += prev_counter;
+        }
+
+        // TODO : This is the bottleneck for moderately dense input arrays.
```

Comment:
@arcondello  @shpface , I could not find a straightforward way of removing this bottleneck. Tried some approaches which work for vectors of primitive elements but not pairs, do you have any idea ? Note I cannot reserve and pushback in parallel in the code below. This is a problem for very large dense matrices. Otherwise fine. With this bottleneck we get 10X speedup on 10 cores for dense matrices. Without would give us (20-40X) speedup though.